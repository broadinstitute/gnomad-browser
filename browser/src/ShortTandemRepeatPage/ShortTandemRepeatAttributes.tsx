import React from 'react'

import AttributeList, { AttributeListItem } from '../AttributeList'
import InlineList from '../InlineList'
import Link from '../Link'

import { ShortTandemRepeat } from './ShortTandemRepeatPage'

type ShortTandemRepeatRepeatUnitsProps = {
  shortTandemRepeat: ShortTandemRepeat
}

const ShortTandemRepeatRepeatUnits = ({ shortTandemRepeat }: ShortTandemRepeatRepeatUnitsProps) => {
  const repeatUnitsByClassification = {}
  shortTandemRepeat.repeat_units.forEach((repeatUnit) => {
    // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
    if (repeatUnitsByClassification[repeatUnit.classification] === undefined) {
      // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
      repeatUnitsByClassification[repeatUnit.classification] = []
    }
    // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
    repeatUnitsByClassification[repeatUnit.classification].push(repeatUnit.repeat_unit)
  })

  if (
    !(repeatUnitsByClassification as any).pathogenic &&
    !(repeatUnitsByClassification as any).benign
  ) {
    return (
      <AttributeListItem
        label={`Repeat unit${(repeatUnitsByClassification as any).unknown.length > 1 ? 's' : ''}`}
      >
        <InlineList
          items={(repeatUnitsByClassification as any).unknown.map((repeatUnit: any) => (
            <span>
              {repeatUnit === shortTandemRepeat.reference_repeat_unit &&
              shortTandemRepeat.repeat_units.length > 1
                ? `${repeatUnit} (reference)`
                : repeatUnit}
            </span>
          ))}
          label={`Repeat unit${(repeatUnitsByClassification as any).unknown.length > 1 ? 's' : ''}`}
        />
      </AttributeListItem>
    )
  }

  return (
    <>
      {(repeatUnitsByClassification as any).pathogenic && (
        <AttributeListItem
          label={`Pathogenic repeat unit${
            (repeatUnitsByClassification as any).pathogenic.length > 1 ? 's' : ''
          }`}
          tooltip="These repeat units have been reported in the literature as pathogenic when they expand beyond a certain threshold."
        >
          <InlineList
            items={(repeatUnitsByClassification as any).pathogenic.map((repeatUnit: any) => (
              <span>
                {repeatUnit === shortTandemRepeat.reference_repeat_unit &&
                shortTandemRepeat.repeat_units.length > 1
                  ? `${repeatUnit} (reference)`
                  : repeatUnit}
              </span>
            ))}
            label={`Pathogenic repeat unit${
              (repeatUnitsByClassification as any).pathogenic.length > 1 ? 's' : ''
            }`}
          />
        </AttributeListItem>
      )}
      {(repeatUnitsByClassification as any).benign && (
        <AttributeListItem
          label={`Benign repeat unit${
            (repeatUnitsByClassification as any).benign.length > 1 ? 's' : ''
          }`}
          tooltip="These repeat units are regarded in the literature as benign, even when expanded."
        >
          <InlineList
            items={(repeatUnitsByClassification as any).benign.map((repeatUnit: any) => (
              <span>
                {repeatUnit === shortTandemRepeat.reference_repeat_unit &&
                shortTandemRepeat.repeat_units.length > 1
                  ? `${repeatUnit} (reference)`
                  : repeatUnit}
              </span>
            ))}
            label={`Benign repeat unit${
              (repeatUnitsByClassification as any).benign.length > 1 ? 's' : ''
            }`}
          />
        </AttributeListItem>
      )}
      {(repeatUnitsByClassification as any).unknown && (
        <AttributeListItem
          label={`Other repeat unit${
            (repeatUnitsByClassification as any).unknown.length > 1 ? 's' : ''
          }`}
          tooltip="These are the other repeat units detected at this locus within gnomAD samples by the call_non_ref_pathogenic_motifs.py script."
        >
          <InlineList
            items={(repeatUnitsByClassification as any).unknown.map((repeatUnit: any) => (
              <span>
                {repeatUnit === shortTandemRepeat.reference_repeat_unit &&
                shortTandemRepeat.repeat_units.length > 1
                  ? `${repeatUnit} (reference)`
                  : repeatUnit}
              </span>
            ))}
            label={`Other repeat unit${
              (repeatUnitsByClassification as any).unknown.length > 1 ? 's' : ''
            }`}
          />
        </AttributeListItem>
      )}
    </>
  )
}

type ShortTandemRepeatAttributesProps = {
  shortTandemRepeat: ShortTandemRepeat
}

const ShortTandemRepeatAttributes = ({ shortTandemRepeat }: ShortTandemRepeatAttributesProps) => {
  return (
    <AttributeList style={{ marginTop: '1.25em' }}>
      <AttributeListItem label="Gene">
        <Link to={`/gene/${shortTandemRepeat.gene.ensembl_id}`}>
          {shortTandemRepeat.gene.symbol}
        </Link>
      </AttributeListItem>
      <AttributeListItem label="Gene region">{shortTandemRepeat.gene.region}</AttributeListItem>
      <AttributeListItem label="Reference region">
        <Link
          to={`/region/${shortTandemRepeat.reference_region.chrom}-${shortTandemRepeat.reference_region.start}-${shortTandemRepeat.reference_region.stop}`}
        >
          {shortTandemRepeat.reference_region.chrom}-{shortTandemRepeat.reference_region.start}-
          {shortTandemRepeat.reference_region.stop}
        </Link>
      </AttributeListItem>
      <ShortTandemRepeatRepeatUnits shortTandemRepeat={shortTandemRepeat} />
    </AttributeList>
  )
}

export default ShortTandemRepeatAttributes
