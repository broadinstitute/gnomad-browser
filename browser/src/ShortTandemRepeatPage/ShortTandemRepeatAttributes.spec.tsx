import React from 'react'
import { expect, test } from '@jest/globals'
import renderer from 'react-test-renderer'
import shortTandemRepeatFactory, { DEFAULT_GENE } from '../__factories__/ShortTandemRepeat'

import ShortTandemRepeatAttributes from './ShortTandemRepeatAttributes'
import { BrowserRouter } from 'react-router-dom'

describe('ShortTandemRepeatAttributes', () => {
  test('has no unexected changes', () => {
    const shortTandemRepeat = shortTandemRepeatFactory.build()
    const tree = renderer.create(
      <BrowserRouter>
        <ShortTandemRepeatAttributes shortTandemRepeat={shortTandemRepeat} />
      </BrowserRouter>
    )
    expect(tree).toMatchSnapshot()
  })

  test('displays no gene link for an STR with empty string for associated gene', () => {
    const shortTandemRepeat = shortTandemRepeatFactory.build({
      gene: { ...DEFAULT_GENE, ensembl_id: '' },
    })
    const tree = renderer.create(
      <BrowserRouter>
        <ShortTandemRepeatAttributes shortTandemRepeat={shortTandemRepeat} />
      </BrowserRouter>
    )
    expect(tree).toMatchSnapshot()
  })
})
