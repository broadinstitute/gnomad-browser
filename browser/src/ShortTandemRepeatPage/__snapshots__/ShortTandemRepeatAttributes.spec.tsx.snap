// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortTandemRepeatAttributes displays no gene link for an STR with empty string for associated gene 1`] = `
<dl
  className="AttributeList-sc-1yjjegy-0 boDRvX"
  style={
    {
      "marginTop": "1.25em",
    }
  }
>
  
  <div>
    <dt>
      Gene region
    </dt>
    <dd>
      coding:polyglutamine
    </dd>
  </div>
  <div>
    <dt>
      Reference region
    </dt>
    <dd>
      <a
        className="Link-sc-14lgydv-0-Link jBvaYQ"
        href="/region/1-10000000-15000000"
        onClick={[Function]}
      >
        1
        :
        10000000
        -
        15000000
      </a>
    </dd>
  </div>
</dl>
`;

exports[`ShortTandemRepeatAttributes has no unexected changes 1`] = `
<dl
  className="AttributeList-sc-1yjjegy-0 boDRvX"
  style={
    {
      "marginTop": "1.25em",
    }
  }
>
  <div>
    <dt>
      Gene
    </dt>
    <dd>
      <a
        className="Link-sc-14lgydv-0-Link jBvaYQ"
        href="/gene/ENSG00000000001"
        onClick={[Function]}
      >
        ABCD1
      </a>
    </dd>
  </div>
  <div>
    <dt>
      Gene region
    </dt>
    <dd>
      coding:polyglutamine
    </dd>
  </div>
  <div>
    <dt>
      Reference region
    </dt>
    <dd>
      <a
        className="Link-sc-14lgydv-0-Link jBvaYQ"
        href="/region/1-10000000-15000000"
        onClick={[Function]}
      >
        1
        :
        10000000
        -
        15000000
      </a>
    </dd>
  </div>
</dl>
`;
